File: project_euler/largest_prime_factor.kt
import kotlin.math.floor
import kotlin.math.sqrt

fun main() {
    print(largestPrimeFactorNaive(13195))
    print(largestPrimeFactorNaive(600851475143))
    print(largestPrimeFactorNaive(2500))
}

fun isPrime(number: Int): Boolean {
    for (i in 2..<number) {
        if (number % i == 0) return false
    }
    return true
}

fun isFactor(limit: Long, i: Int): Boolean {
    return  (limit % i).toInt() == 0
}

fun largestPrimeFactorNaive(limit: Long): Int {
    val factors = mutableListOf<Int>();
    val squareRootOfLimit = (floor(sqrt(limit.toDouble())) - 2).toInt()
    for (i in 2..<squareRootOfLimit) {
        if (isPrime(i) && isFactor(limit, i)) {
            factors.add(i);
            val complementaryFactor = floor((limit / i).toDouble()).toInt();
            if (isPrime(complementaryFactor)) factors.add(complementaryFactor)
        }
    }
    return factors.removeLast()
}

File: project_euler/largest_prime_factor.ts
const isPrime = (number: number): boolean => {
	for (let i = 2; i < number; i++) {
		if (number % i == 0) return false;
	}
	return true;
}

const isFactor = (limit: number, i: number): boolean => {
	return limit % i == 0;
}

// https://www.grae.io/post/euler_problem_3/
const largestPrimeFactorNaive = (limit: number): number => {
	let factors: number[] = [];
	for (let i = 2; i < Math.floor(Math.sqrt(limit)); i++) {
		if (isPrime(i) && isFactor(limit, i)) {
			factors = [...factors, i];
			const complementaryFactor = Math.floor(limit / i);
			if (isPrime(complementaryFactor)) {
				factors = [...factors, complementaryFactor];
			}
		}
	}
	return factors[factors.length - 1];
}

console.log(largestPrimeFactorNaive(13195));
console.log(largestPrimeFactorNaive(600851475143));
console.log(largestPrimeFactorNaive(2500));